// eways-backend/config/email.js
/**
 * Email Configuration
 * Nodemailer setup for email services
 * 
 * @author Eways Team
 */

const nodemailer = require('nodemailer');

/**
 * Create email transporter based on environment configuration
 * @returns {Object} Nodemailer transporter
 */
const createTransporter = () => {
  const config = {
    service: process.env.EMAIL_SERVICE || 'gmail',
    host: process.env.EMAIL_HOST || 'smtp.gmail.com',
    port: parseInt(process.env.EMAIL_PORT) || 587,
    secure: false, // true for 465, false for other ports
    auth: {
      user: process.env.EMAIL_USER,
      pass: process.env.EMAIL_PASS,
    },
    tls: {
      rejectUnauthorized: false
    }
  };

  return nodemailer.createTransport(config);
};

/**
 * Email templates
 */
const emailTemplates = {
  // Contact form notification to admin
  contactNotification: (contactData) => ({
    subject: `New Contact Form Submission - ${contactData.subject}`,
    html: `
      <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
        <h2 style="color: #333; border-bottom: 2px solid #007bff; padding-bottom: 10px;">
          New Contact Form Submission
        </h2>
        
        <div style="background-color: #f8f9fa; padding: 20px; border-radius: 5px; margin: 20px 0;">
          <h3 style="color: #495057; margin-top: 0;">Contact Details:</h3>
          <p><strong>Name:</strong> ${contactData.name}</p>
          <p><strong>Email:</strong> ${contactData.email}</p>
          <p><strong>Phone:</strong> ${contactData.phone || 'Not provided'}</p>
          <p><strong>Subject:</strong> ${contactData.subject}</p>
        </div>
        
        <div style="background-color: #fff; padding: 20px; border-left: 4px solid #007bff; margin: 20px 0;">
          <h3 style="color: #495057; margin-top: 0;">Message:</h3>
          <p style="line-height: 1.6;">${contactData.message}</p>
        </div>
        
        <div style="margin-top: 30px; padding: 15px; background-color: #e9ecef; border-radius: 5px;">
          <small style="color: #6c757d;">
            <strong>Submission Details:</strong><br>
            Time: ${new Date().toLocaleString()}<br>
            IP Address: ${contactData.ip_address || 'Unknown'}<br>
            User Agent: ${contactData.user_agent || 'Unknown'}
          </small>
        </div>
        
        <div style="margin-top: 30px; text-align: center;">
          <p style="color: #6c757d; font-size: 12px;">
            This email was automatically generated by the Eways contact form system.
          </p>
        </div>
      </div>
    `
  }),

  // Auto-reply to contact form submitter
  contactAutoReply: (contactData) => ({
    subject: 'Thank you for contacting Eways - We\'ll be in touch soon!',
    html: `
      <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
        <div style="text-align: center; margin-bottom: 30px;">
          <h1 style="color: #007bff;">Eways</h1>
        </div>
        
        <h2 style="color: #333;">Thank you for reaching out!</h2>
        
        <p>Dear ${contactData.name},</p>
        
        <p>We have received your message and appreciate you taking the time to contact us. Our team will review your inquiry and get back to you within 24-48 hours.</p>
        
        <div style="background-color: #f8f9fa; padding: 20px; border-radius: 5px; margin: 20px 0;">
          <h3 style="color: #495057; margin-top: 0;">Your Message Summary:</h3>
          <p><strong>Subject:</strong> ${contactData.subject}</p>
          <p><strong>Message:</strong> ${contactData.message.substring(0, 200)}${contactData.message.length > 200 ? '...' : ''}</p>
        </div>
        
        <p>In the meantime, feel free to explore our website to learn more about our services and projects.</p>
        
        <div style="margin: 30px 0; text-align: center;">
          <a href="${process.env.FRONTEND_URL}" style="background-color: #007bff; color: white; padding: 12px 30px; text-decoration: none; border-radius: 5px; display: inline-block;">
            Visit Our Website
          </a>
        </div>
        
        <p>Best regards,<br>The Eways Team</p>
        
        <hr style="margin: 30px 0; border: none; border-top: 1px solid #e9ecef;">
        
        <div style="text-align: center;">
          <p style="color: #6c757d; font-size: 12px;">
            This is an automated response. Please do not reply to this email.<br>
            For urgent matters, please call us directly or visit our website.
          </p>
        </div>
      </div>
    `
  }),

  // Admin login notification
  adminLoginNotification: (adminData) => ({
    subject: 'Admin Login Notification - Eways Backend',
    html: `
      <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
        <h2 style="color: #333;">Admin Login Notification</h2>
        
        <p>An admin user has logged into the Eways backend system.</p>
        
        <div style="background-color: #f8f9fa; padding: 20px; border-radius: 5px; margin: 20px 0;">
          <p><strong>Username:</strong> ${adminData.username}</p>
          <p><strong>Email:</strong> ${adminData.email}</p>
          <p><strong>Login Time:</strong> ${new Date().toLocaleString()}</p>
          <p><strong>IP Address:</strong> ${adminData.ip_address || 'Unknown'}</p>
        </div>
        
        <p style="color: #dc3545; font-size: 14px;">
          If this was not you, please secure your account immediately.
        </p>
      </div>
    `
  })
};

/**
 * Send email using configured transporter
 * @param {Object} emailOptions - Email options (to, subject, html, etc.)
 * @returns {Promise} Send email promise
 */
const sendEmail = async (emailOptions) => {
  try {
    // For development, just log emails instead of sending
    if (process.env.NODE_ENV === 'development') {
      console.log('\n📧 ===== EMAIL SIMULATION =====');
      console.log('📤 TO:', emailOptions.to);
      console.log('📋 SUBJECT:', emailOptions.subject);
      console.log('📝 HTML CONTENT:');
      console.log(emailOptions.html?.substring(0, 500) + '...' || 'No HTML content');
      console.log('===============================\n');
      
      return { 
        messageId: 'simulated-' + Date.now(),
        response: 'Email logged to console (development mode)' 
      };
    }

    const transporter = createTransporter();
    
    const mailOptions = {
      from: process.env.EMAIL_FROM || process.env.EMAIL_USER,
      to: emailOptions.to,
      subject: emailOptions.subject,
      html: emailOptions.html,
      text: emailOptions.text,
      ...emailOptions
    };

    const result = await transporter.sendMail(mailOptions);
    console.log('✅ Email sent successfully:', result.messageId);
    return result;
  } catch (error) {
    console.error('❌ Email sending failed:', error.message);
    throw error;
  }
};

/**
 * Test email configuration
 * @returns {Promise} Test result
 */
const testEmailConfig = async () => {
  try {
    const transporter = createTransporter();
    await transporter.verify();
    console.log('✅ Email configuration is valid');
    return true;
  } catch (error) {
    console.error('❌ Email configuration test failed:', error.message);
    return false;
  }
};

module.exports = {
  createTransporter,
  emailTemplates,
  sendEmail,
  testEmailConfig
};
